### PR TITLE
web4: scroll comments into view

### PR DESCRIPTION
**Pull request by petersalomonsen**
Enables "scroll into view" for proposal comments functionality for https://github.com/NEAR-DevHub/neardevhub-bos/pull/868

Example link to comment being scrolled into view: https://devhublink.testnet.page/proposal/127?accountId=theori.near&blockHeight=121684702

### CODE REVIEW

**Comment by petersalomonsen** [comment]
@race-of-slots please, include my PR in the race

**Comment by petersalomonsen** [comment]
@race-of-sloths  please, include my PR in the race

**Comment by race-of-sloths** [comment]
@petersalomonsen Thank you for your contribution! Your pull request is now a part of the [Race of Sloths](https://race-of-sloths.com)!
New Sloth joined the Race! Welcome!

[<picture>
    <source media="(prefers-color-scheme: dark)" srcset="https://badge.race-of-sloths.com/petersalomonsen?type=bot&theme=dark">
    <source media="(prefers-color-scheme: light)" srcset="https://badge.race-of-sloths.com/petersalomonsen?type=bot&theme=light">
    Shows profile picture for the author of the PR
</picture>](https://race-of-sloths.com/profile/petersalomonsen)




<details><summary>Current status: <i>executed</i></summary>

| Reviewer | Score |
|--------|--------|
| @frol  | 3 |

**The average score is 3**

@petersalomonsen check out your results on the [Race of Sloths Leaderboard!](https://race-of-sloths.com/leaderboard) and in the [profile](https://race-of-sloths.com/profile/petersalomonsen)

</details><details><summary>What is the Race of Sloths</summary>
<p>
Race of Sloths is a friendly competition where you can participate in challenges and compete with other open-source contributors within your normal workflow

For contributors:

- Tag @race-of-sloths inside your pull requests
- Wait for the maintainer to review and score your pull request
- Check out your position in the [Leaderboard](https://race-of-sloths.com/leaderboard)
- Keep weekly and monthly streaks to reach higher positions
- Boast your contributions with a dynamic picture of your [Profile](https://race-of-sloths.com/profile/petersalomonsen)

For maintainers:
- Score pull requests that participate in the Race of Sloths
- Engage contributors with fair scoring and fast responses so they keep their streaks
- Promote the Race to the point where the Race starts promoting you
- Grow the community of your contributors

Feel free to check [our website](https://race-of-sloths.com) for additional details!
</p>
</details> 

<details>
<summary>Bot commands</summary>

- For contributors
  - **Include a PR:** `@race-of-sloths include` to enter the Race with your PR
- For maintainers:
  - **Assign points:** `@race-of-sloths score [1/2/3/5/8/13]` to award points based on your assessment.
  - **Reject this PR:** `@race-of-sloths exclude` to send this PR back to the drawing board.
  - **Exclude repo:** `@race-of-sloths pause` to stop bot activity in this repo until `@race-of-sloths unpause` command is called

</details>

File: src/web4/scroll_comment_into_view.js
Comment (frol) [inline src/web4/scroll_comment_into_view.js:6]: What is this `for` loop for?

**Review by frol — COMMENTED** [review]
@petersalomonsen Can you provide an example link to test?

**Comment by petersalomonsen** [comment]
> @petersalomonsen Can you provide an example link to test?

yeah, you can also find it in the frontend PR as referred to in the description, but here is a link the deployment on testnet ( with a frontend that points to mainnet ): https://devhublink.testnet.page/proposal/127?accountId=theori.near&blockHeight=121684702

File: src/web4/scroll_comment_into_view.js
Comment (petersalomonsen) [reply src/web4/scroll_comment_into_view.js:6]: The `for` loop is for making up to 20 attempts every half second to see if the comment has become visible ( it does takes some time ), and if that is the case it will be scrolled into the view.

I thought it may be reasonable to make a finite number of attempts, in case the `accountId` and `blockHeight` parameters exists, but the page does not have the element id that can be scrolled to. This is the case for other pages than proposals.

**Comment by Tguntenaar** [comment]
The preview does not work for me:

**Review by Tguntenaar — COMMENTED** [review]
@frol @petersalomonsen quick question:

For the [tracking of referral sources](https://github.com/NEAR-DevHub/neardevhub-bos/issues/879) we need to have access to the PostHog API key inside the javascript. I think the best way is to [expose the env variable](https://doc.rust-lang.org/book/ch12-05-working-with-environment-variables.html) to the rust build process and replace part of the imported javascript string.

Do you guys agree?

**Comment by frol** [comment]
@Tguntenaar I find it not the best solution:

1. It makes the contract build "not reproducible" without the knowledge to the API key (cargo-near reproducible builds do not even support env variables passing, so I would avoid that)
2. There is too much of indirection for this API key to travel from github build process of a contract down to web4 gateway JS code
3. We already have the key in our widgets
4. @petersalomonsen Can we pass extra variable to the BOS components via `context` (next to `context.networkId`) or maybe at least via `props` (e.g. implicitly passing `?gateway=devhub.near.page` when rendering the top-level component)?

**Comment by petersalomonsen** [comment]
@Tguntenaar the preview link works now ( it was pointing to devgovgigs.petersalomonsen.near preview environment which I have deleted, now it is pointing to devhub.near )

@frol @Tguntenaar Regarding tracking the gateway being used. We don't need to include it in the contract build process.

There are other alternatives, and the easiest is to pass the additional information we want to track into `props`. Then it will be automatically visible in PostHOG too.

Passing into `context` seems not to be supported out of the box in the [VM](https://github.com/NearSocial/VM/blob/6047c6a9b96f3de14e600c1d2b96c432bbb76dd4/src/lib/components/Widget.js#L195), so I think `props` is better then.

We can also go into more "hacky" alternatives like extracting the key from the widget, but I would prefer just passing the information we want to track into `props`.


And finally, regarding this PR. The solution on looking for an element id to scroll to, is an attempt to create a generic solution for finding such elements that can be scrolled to. An alternative would be to lock it to only proposals, and only then wait for the element to appear. In that case we need to revisit this every time we want to support scrolling to a comment element for something else than proposals, so I would prefer the current solution @frol .

**Comment by Tguntenaar** [comment]
Thanks for the quick feedback guys, @petersalomonsen the preview looks good to me!

Just for reference the issue we are talking about [#879](https://github.com/NEAR-DevHub/neardevhub-bos/issues/879)

**Review by Tguntenaar — APPROVED** [review]
LGTM

**Review by frol — APPROVED** [review]
@race-of-sloths score 3

**Comment by race-of-sloths** [comment]
### 🌟 Score recorded!

@frol, thank you for scoring this pull request in the Race of Sloths!

**Comment by race-of-sloths** [comment]
### ✅ PR is finalized! 
Your contribution is much appreciated with a final score of 3!
You have received 45 (30 base + 15 weekly bonus) Sloth points for this contribution


Another weekly streak completed, well done @petersalomonsen! To keep your weekly streak and get another bonus make pull request next week! Looking forward to see you in race-of-sloths

**Pull request merged**
